### PR TITLE
Use all __future__ imports in beets core

### PR DESCRIPTION
**Pull request by brunal**
This PR make every module under `beets` (and related test modules) use all `__future__` imports (division, print_function, absolute_import, and the infamous `unicode_literals). As written in a commit message the long-term goal is python 3 compatibility.

This refactorization was mostly test-driven. Given than the tests cover 75% of the code base, some things may be broken. The process was
    1. Include the 3 easy `__future__` imports
    2. Package by package add the `unicode_literals` import. Also add it to the respective tests.

I can so far say that most problematic package for python 3 conversion will probably be `beets.util.functtemplate` (API changes to the AST) and `beets.mediafile` (low-level manipulations).
Moreover, I discovered that tests for the `convert` plugin were tremendously useful and were the only one to test the external, global behaviour of beets.
Finally, this PR only covers the `beets` package and not the `beetsplug` one. I plan on doing the same for the latter but lack of unit tests on several plugins dampened my progress.

edit:

* using pathlib would probably be a *big* relief over all path management statement in the codebase
* there are way too many `isinstance` statements in many places, clarifying the data models seems necessary before going further

### CODE REVIEW

**Comment by sampsyo** [comment]
Wow, thank you for tackling yet another huge refactoring task! I'm happy to be moving closer to Python 3 status.

A few assorted notes:

* How did `unicode_literals` go? Was anything unexpected? The typical objection to using `unicode_literals` for portable code is that there are certain Python APIs that expect "native" strings across the versions—meaning that, on Python 2, they deal in bytestrings, but on Python 3, they want Unicode strings. A canonical example is urllib (although that's not used in core). I *suspect* (although don't know) that SQLite row names might have this same "native string" behavior. Did anything like that come up? The rest of these are obvious wins; unicode_literals is somewhat more ambiguous. (That said, all these explicit `b''` literals are *great* for clarifying intent regardless.)
* May I humbly suggest a run through [unicode-nazi](https://github.com/mitsuhiko/unicode-nazi) to check for any unintended conversions?
* I feel fine leaving some plugins un-ported until we have tests for them. Tests then refactoring. I suppose this is the downside to shipping plugins with core—we actually have to take care of them! :smiley:
* I see the problem with `functemplate`. I think we should consider throwing away my ridiculous AST-generation compiler, which has never been much of a performance win.
* Yes, MediaFile will be difficult. We will need to be very careful with all our strings.
* Pathlib would definitely be interesting to explore (c.f. #621). This will need to be done with **tremendous** care, however—path encodings are an incredibly insidious source of many platform-specific bugs, and it's not yet clear how "right" pathlib is about dealing with encodings.
* Got specific examples about the `isinstance` tests? I agree we do this too often but I don't have a catalog of the offenses.

**Comment by brunal** [comment]
`unicode_literals`: overall it was not that hard -- doing it file-by-file was sure easier than my first try when I converted all of them with a `sed`! In some cases I was not sure where to encode/decode the object: the desired API is not always easy to guess. Overall I found there were to be bytes in 4 places:
- AST identifiers & co.
- paths and all that relates to them (so their db manipulations as well)
- metadata embedded in files (so `mediafiles`)
- db row names (it explicitly fails with unicode)
The trickiest parts were some tests who would randomly fail with a low probability -- this was due to _ and % escaping in `beets.library`, and the tests would only fail when the temp folder name had a _!

unicode-nazi reports lots of warnings, I'll fix them. I should also replace *all* mentions of `str` with `bytes`.

Many functions check for the string type when I believe they're always receiving the same thing/the data model should be clarified. See `confit`, `mediafile`, `library` `ui` modules: `ui.print_()` `ui._colordiff()`, path management in library, ...

**Comment by brunal** [comment]
Unicodenazi allowed me to catch numerous automatic conversions, now solved. The only remaining ones seem to happen on `OptionParser` constructor call but I don't think they mean anything.
Note that I only looked at warnings happening in `beets` and not in other modules, for they're too numerous and often unrelated to beets.

To fix all those errors making all plugins use `unicode_literals` was the simplest solution, which I did.

File: beetsplug/convert.py [hunk 1]
```diff
@@ -175,8 +178,8 @@ def encode(self, command, source, dest, pretend=False):
         args = shlex.split(command)
         for i, arg in enumerate(args):
             args[i] = Template(arg).safe_substitute({
-                'source': source,
-                'dest': dest,
+                'source': source.decode('utf8'),
+                'dest': dest.decode('utf8'),
```
Comment (sampsyo) [inline beetsplug/convert.py]: Note to self: Need to come back and test this. I would have expected we were constructing bytestrings for the command. Does a unicode command work?

**Comment by sampsyo** [comment]
Yay! Sorry this took me so long to process.

**Comment by brunal** [comment]
Thank you for merging this :-) no problem, it was quite huge. I just hope it does not break any functionality in a subtle way on esoteric platforms!

**Pull request merged**